homeassistant/components/switch/rainmachine.py

```diff
@@ -1,208 +1,400 @@
+        """Calculate running days from an RM string ("0010001100")."""
+        freq_type = self._obj['frequency']['type']
+        freq_param = self._obj['frequency']['param']
+        if freq_type == 0:
+            return 'Daily'
+
+        if freq_type == 1:
+            return 'Every {0} Days'.format(freq_param)
+
+        if freq_type == 2:
+            return ', '.join([
+                DAYS[idx] for idx, val in enumerate(freq_param[2:-1][::-1])
+                if val == '1'
+            ])
+
+        if freq_type == 4:
+            return '{0} Days'.format('Odd' if freq_param == '1' else 'Even')
+
+        return None
 
     def turn_off(self, **kwargs) -> None:
         """Turn the program off."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._client.programs.stop(self.rainmachine_entity_id)
-        except exceptions.BrokenAPICall:
-            _LOGGER.error('programs.stop currently broken in remote API')
-        except exceptions.HTTPError as exc_info:
+            self.rainmachine.client.programs.stop(self._rainmachine_entity_id)
+            dispatcher_send(self.hass, PROGRAM_UPDATE_TOPIC)
+        except HTTPError as exc_info:
             _LOGGER.error('Unable to turn off program "%s"', self.unique_id)
             _LOGGER.debug(exc_info)
 
     def turn_on(self, **kwargs) -> None:
         """Turn the program on."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._client.programs.start(self.rainmachine_entity_id)
-        except exceptions.BrokenAPICall:
-            _LOGGER.error('programs.start currently broken in remote API')
-        except exceptions.HTTPError as exc_info:
+            self.rainmachine.client.programs.start(self._rainmachine_entity_id)
+            dispatcher_send(self.hass, PROGRAM_UPDATE_TOPIC)
+        except HTTPError as exc_info:
             _LOGGER.error('Unable to turn on program "%s"', self.unique_id)
             _LOGGER.debug(exc_info)
 
-    @Throttle(MIN_SCAN_TIME, MIN_SCAN_TIME_FORCED)
     def update(self) -> None:
         """Update info for the program."""
-        import regenmaschine.exceptions as exceptions
+        from regenmaschine.exceptions import HTTPError
 
         try:
-            self._entity_json = self._client.programs.get(
-                self.rainmachine_entity_id)
-        except exceptions.HTTPError as exc_info:
+            self._obj = self.rainmachine.client.programs.get(
+                self._rainmachine_entity_id)
+
+            self._attrs.update({
+                ATTR_CS_ON: self._obj.get('cs_on'),
+                ATTR_CYCLES: self._obj.get('cycles'),
+                ATTR_DELAY: self._obj.get('delay'),
+                ATTR_DELAY_ON: self._obj.get('delay_on'),
+                ATTR_FREQUENCY: self._calculate_running_days(),
```

Comment:
We should never have relative time in the attributes.

Reply:
10-4.